src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,594 @@
+            # handle case like:
+            # session.udf.register(
+            #    lambda x, y: x + y, ...
+            # )
+            code_text = f"wrapper({code_text})"
+        elif "unmatched ')'" in str(exc):
+            # handle case like:
+            # session.udf.register(
+            #    lambda x, y: x + y, ...)
+            code_text = f"wrapper({code_text}"
+        source_ast = ast.parse(code_text)
+    lambda_node = next(
+        (node for node in ast.walk(source_ast) if isinstance(node, ast.Lambda)), None
+    )
+    if not lambda_node:
+        raise TypeError("lambda function can not be extracted")
+
+    lines = code_text.splitlines()
+    # single line lambda
+    if len(lines) == 1:
+        return code_text[lambda_node.col_offset : lambda_node.end_col_offset]
+
+    lambda_code_text = ""
+    # lambda of multiple lines
+    # handle case like:
+    # session.udf.register(
+    #    lambda x, y:\
+    #    x + y, ...)
+    for line_idx in range(lambda_node.lineno - 1, lambda_node.end_lineno):
+        line = lines[line_idx]
+        if line_idx == 0:
+            lambda_code_text = f"{lambda_code_text}{line[lambda_node.col_offset:]}\n"
+        elif line_idx == lambda_node.end_lineno - 1:
+            lambda_code_text = f"{lambda_code_text}{line[: lambda_node.end_col_offset]}"
+        else:
+            lambda_code_text = f"{lambda_code_text}{line}\n"
+    return lambda_code_text.strip()
+
+
+def extract_submodule_imports(
+    func: FunctionType, top_level_modules: Iterable[ModuleType]
+) -> Set[Tuple[str, str]]:
+    """
+    Get submodule imports, the func code co_names only gives the top level module names, the submodule imports
+    have to be inferred manually. Consider the following example:
+
+    import a1.a2.a3.a4
+    def func():
+        a1.a2.a3.a4.foo()
+
+    func.__code__.co_names only contains ("a1", "a2", "a3", "a4", "foo") which does not include the
+    complete import path information.
+
+    To reconstruct "a1.a2.a3.a4", the current strategy is to import each prefix import of the import chains.
+    This is not a perfect solution as we could import modules not used, but it works.
+
+    Args:
+        func: The target function to generate source code for.
+        top_level_modules: The name of top level modules from which to search the referenced imported objects
+            or submodules.
+
+    Returns:
+        A set of tuple with each tuple composed of two string, the first one is actual name for the imported object,
+        and the second one is alias.
+
+    """
+    imports = set()
+    for module in top_level_modules:
+        module_prefix = f"{module.__name__}."
+        for name in [m for m in sys.modules if m.startswith(module_prefix)]:
+            tokens = set(name[len(module_prefix) :].split("."))
+            if not tokens - set(
+                func.__code__.co_names
+            ):  # only add imports that co_names contains
+                imports.add((name, name))
```

Comment:
the docstring says the second item of the tuple is the alias, but from here actually they are the same?

Reply:
it could be an alias, I've taken Shixuan's idea to use named tuple with default values. it should now look more intuitive.